src/aspire/source/micrograph.py

```diff
@@ -0,0 +1,299 @@
+            raise ValueError(
+                "The micrograph size must be larger or equal to the simulation's image size."
+            )
+
+        self.particle_box_size = simulation.L
+        self.particles_per_micrograph = particles_per_micrograph
+        self.total_particle_count = (
+            self.micrograph_count * self.particles_per_micrograph
+        )
+        self.dtype = simulation.dtype
+
+        self.noise_adder = noise_adder
+
+        if boundary is None:
+            self.boundary = self.particle_box_size // 2
+        else:
+            if (
+                boundary < (0 - self.particle_box_size // 2)
+                or boundary > self.micrograph_size // 2
+            ):
+                raise RuntimeError("Illegal boundary value.")
+            self.boundary = boundary
+
+        if interparticle_distance is None:
+            self.interparticle_distance = self.particle_box_size
+        else:
+            self.interparticle_distance = interparticle_distance
+
+        # Create the radial mask for each center
+        g2d = grid_2d(int(2 * self.interparticle_distance), normalized=False)
+        radial_mask = g2d["r"] <= self.interparticle_distance
+        self.grid_x = g2d["x"].astype(int)[radial_mask]
+        self.grid_y = g2d["y"].astype(int)[radial_mask]
+
+        # Calculate the proper padding and create the mask
+        self.pad = int(max(self.particle_box_size, self.interparticle_distance))
+        self.mask = np.full(
+            (
+                self.micrograph_count,
+                int(self.micrograph_size + 2 * self.pad),
+                int(self.micrograph_size + 2 * self.pad),
+            ),
+            False,
+            dtype=bool,
+        )
+        self.mask_boundary = self.pad + self.boundary
+        self.mask[
+            :,
+            self.mask_boundary : -1 * self.mask_boundary,
+            self.mask_boundary : -1 * self.mask_boundary,
+        ] = True
+
+        # Create the centers
+        self.centers = np.zeros(
+            (self.micrograph_count, self.particles_per_micrograph, 2), dtype=int
+        )
+        self.fail_limit = pass_threshold * self.micrograph_count
+        self.fail_count = 0
+        for i in range(self.micrograph_count):
+            self.centers[i] = self._create_centers(i)
+
+        self._clean_micrographs_accessor = _ImageAccessor(
+            self._clean_micrographs, self.micrograph_count
+        )
+        self._micrographs_accessor = _ImageAccessor(
+            self._micrographs, self.micrograph_count
+        )
+
+        self.images = _ImageAccessor(self._images, self.total_particle_count)
+
+    def _create_centers(self, micrograph):
+        """
+        Creates centers for the given micrograph if the fail threshold isn't met.
+
+        param micrograph: The ID of the micrograph.
+        return: A numpy array containing the generated centers.
+        """
+        while self.fail_count < self.fail_limit:
```

Comment:
We need to talk about fail_count (and resetting it) etc.  It probably doesn't work well if we repeatedly ask for micrographs...

Reply:
Discussed, this seems to only happen once.  Let's make them private and all good!

My mistake :).